examples/cosmos/ics20/README.md

```diff
@@ -1,20 +1,164 @@
+        destChannel: "channelToA",
+        destPort: "transfer",
+        sourceChannel: "channelToB",
+        sourcePort: "transfer"
+      }
+    ),
+}
+```
+
+Also note that the outgoing IBC packet has been added to the chain state.
+
+By executing the `sendPacket` action, the packet was only sent, but not yet relayed/received. We can convince ourselves by checking that the bank state on chain `"B"` has not yet changed:
+
+```bluespec
+>>> chainStates.get("B").bank
+Map()
+```
+
+Let's now take action `receivePacket`, which triggers receival of the packet:
+
+```bluespec
+>>> receivePacket("A", "B")
+true
+>>> chainStates.get("B")
+{
+  bank: Map("bob" -> Map({ baseDenom: "atom", path: [{ channel: "channelToA", port: "transfer" }] } -> 1)),
+  channelEscrowAddresses: Map("channelToA" -> "escrow_account", "channelToC" -> "escrow_account"),
+  channels: Map("channelToA" -> "channelToB", "channelToC" -> "channelToB"),
+  inAcknowledgements: Set(),
+  inTimeouts: Set(),
+  outPackets: Set(),
+  receivedButUnacknowledgedPackets: Set()
+}
+```
+
+Now Bob has been minted the sent token. We can also see that the token's base denomination has been extended with a trace of its origin (`{ channel: "channelToA", port: "transfer" }`).
+
+Receival of the packet has created an IBC acknowledgement. Although receival of a successful acknowledgement has no effect, let's receive it on chain `"A"`:
+
+```bluespec
+>>> chainStates.get("A").inAcknowledgements
+Set({ errorMessage: "", success: true })
+>>> receiveAck("A")
+true
+```
 
-- `bank` contains an abstract bank module that supports minting, burning, and
-  transfer of fungible tokens.
+To take the protocol through a complete transfer, in a single step, use the `sendTransfer` run:
+
+```bluespec
+>>> sendTransfer("A", "B", ATOM, 50, "alice", "charlie")
+true
+>>> chainStates.get("A").bank
+Map(
+  "alice" -> Map({ baseDenom: "atom", path: [] } -> 49),
+  "escrow_account" -> Map({ baseDenom: "atom", path: [] } -> 51)
+)
+>>> chainStates.get("B").bank
+Map(
+  "bob" -> Map({ baseDenom: "atom", path: [{ channel: "channelToA", port: "transfer" }] } -> 1),
+  "charlie" -> Map({ baseDenom: "atom", path: [{ channel: "channelToA", port: "transfer" }] } -> 50)
+)
+```
+
+We also provide actions `timeoutPacket` and `receiveTimeout`, that can be called instead of `receivePacket` and `receiveAck` to simulate a timed out transfer.
+
+### Tests
+
+Instead of exploring the protocol interactively, we can encode longer protocol runs, together with assertions that assure correct behavior, as Quint tests. Use the Quint test runner to execute the tests in [`ics20.qnt`](./ics20.qnt):
+
+```sh
+$ quint test --main=ics20Test examples/cosmos/ics20/ics20.qnt
+
+  ics20Test
+    ok ABCACBATest passed 10000 test(s)
+    ok TransferTest passed 10000 test(s)
+    ok FailedAckTest passed 10000 test(s)
+    ok TimeoutTest passed 10000 test(s)
```

Comment:
is there non-determinism inside? I wonder why the engine is running 10000 tests

Reply:
what constitutes non-determinism?
there's `oneOf` for picking a packet, but in the test the set it is picking out of always has cardinality 1.